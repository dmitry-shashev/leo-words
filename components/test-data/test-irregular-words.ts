import { IrregularWord } from '@/models/irregular-word'

export const testIrregularWords: Array<IrregularWord> = [
  {
    id: 1,
    infinitive: 'bite',
    pastSimple: 'bit',
    pastParticiple: 'bitten',
  },
  {
    id: 2,
    infinitive: 'draw',
    pastSimple: 'drew',
    pastParticiple: 'drawn',
  },
  {
    id: 3,
    infinitive: 'do',
    pastSimple: 'did',
    pastParticiple: 'done',
  },
  {
    id: 4,
    infinitive: 'fly',
    pastSimple: 'flew',
    pastParticiple: 'flown',
  },
  {
    id: 5,
    infinitive: 'drive',
    pastSimple: 'drove',
    pastParticiple: 'driven',
  },
]
