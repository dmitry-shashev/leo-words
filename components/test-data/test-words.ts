import { Word } from '@/models/word'

export const testWords: Array<Word> = [
  {
    id: 31684,
    combinedTranslation: 'проповедовать',
    wordValue: 'preach',
    transcription: 'prˈiːtʃ',
    picture: 'https://contentcdn.lingualeo.com/uploads/picture/634277.png',
    pronunciation: 'https://audiocdn.lingualeo.com/v2/3/cHJlYWNo6ba93e74.mp3',
    created: 1433462400,
    added: 1433462400,
  },
  {
    id: 35298,
    combinedTranslation: 'польза',
    wordValue: 'sake',
    transcription: 'seɪk',
    picture: 'https://contentcdn.lingualeo.com/uploads/picture/326972.png',
    pronunciation: 'https://audiocdn.lingualeo.com/v2/3/c2FrZQ==30ec034a.mp3',
    created: 1433462400,
    added: 1433462400,
  },
  {
    id: 39328,
    combinedTranslation: 'прогулка',
    wordValue: 'stroll',
    transcription: 'stroʊl',
    picture: 'https://contentcdn.lingualeo.com/uploads/picture/2898240.png',
    pronunciation: 'https://audiocdn.lingualeo.com/v2/3/c3Ryb2xsf9e1435d.mp3',
    created: 1433462400,
    added: 1433462400,
  },
  {
    id: 43022,
    combinedTranslation: 'предпринимать',
    wordValue: 'undertake',
    transcription: 'ˈʌndərˌteɪk',
    picture: 'https://contentcdn.lingualeo.com/uploads/picture/19282.png',
    pronunciation:
      'https://audiocdn.lingualeo.com/v2/3/dW5kZXJ0YWtlce0a68dd.mp3',
    created: 1433462400,
    added: 1433462400,
  },
  {
    id: 54653,
    combinedTranslation: 'эффективно',
    wordValue: 'effectively',
    transcription: 'ɪˈfɛktɪvli',
    picture: 'https://contentcdn.lingualeo.com/uploads/picture/31663.png',
    pronunciation:
      'https://audiocdn.lingualeo.com/v2/3/ZWZmZWN0aXZlbHk=c9b5216a.mp3',
    created: 1433462400,
    added: 1433462400,
  },
]
