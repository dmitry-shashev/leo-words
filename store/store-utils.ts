import { Word } from '@/models/word'
import { IrregularWord } from '@/models/irregular-word'

export function getEmptyWord(): Word {
  return {
    id: 0,
    wordValue: '',
    pronunciation: '',
    picture: '',
    combinedTranslation: '',
    transcription: '',
    created: 0,
  }
}

export function getEmptyIrregularWord(): IrregularWord {
  return {
    id: 0,
    infinitive: '',
    pastSimple: '',
    pastParticiple: '',
  }
}
