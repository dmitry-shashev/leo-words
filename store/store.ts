import {
  combineReducers,
  configureStore,
  EnhancedStore,
} from '@reduxjs/toolkit'
import settingsReducer from './slices/settingsSlice'
import wordsReducer from './slices/wordsSlice'
import irregularWordsReducer from './slices/irregularWordsSlice'
import { useDispatch } from 'react-redux'
import storage from 'redux-persist/lib/storage'
import {
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
  persistStore,
  persistReducer,
} from 'redux-persist'

const persistConfig = {
  key: 'root',
  storage,
}

const reducer = combineReducers({
  settingsReducer,
  wordsReducer,
  irregularWordsReducer,
})

const persistedReducer = persistReducer(persistConfig, reducer)

export function createStore(): EnhancedStore {
  return configureStore({
    reducer: persistedReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: {
          ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
        },
      }),
  })
}

const store = createStore()
let persistor = persistStore(store)

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export const useAppDispatch: () => AppDispatch = useDispatch

export { store, persistor }
