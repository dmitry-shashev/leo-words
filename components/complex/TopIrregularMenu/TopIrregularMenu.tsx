'use client'

import { FC } from 'react'
import { useSelector } from 'react-redux'
import { useAppDispatch } from '@/store/store'
import {
  getAllIrregularWords,
  getFinishedIrregularWords,
  getIsByIrregularWrongWords,
  getWrongIrregularWords,
  setAllIrregularWords,
  toggleByWrongIrregularWords,
} from '@/store/slices/irregularWordsSlice'
import { SquareBtn } from '@/components/simple/SquareBtn/SquareBtn'
import { WordsBtn } from '@/components/simple/WordsBtn/WordsBtn'
import { shuffleArr } from '@/utils/shuffleArr'

export const TopIrregularMenu: FC = () => {
  const dispatch = useAppDispatch()

  const allWords = useSelector(getAllIrregularWords)
  const isByWrongIrregularWords = useSelector(getIsByIrregularWrongWords)
  const wrongIrregularWords = useSelector(getWrongIrregularWords)
  const finishedIrregularWords = useSelector(getFinishedIrregularWords)

  const onWrongWords = (): void => {
    dispatch(toggleByWrongIrregularWords())
  }

  const onShuffle = (): void => {
    const newWords = shuffleArr(allWords)
    dispatch(setAllIrregularWords(newWords))
  }

  return (
    <div className="flex flex-row justify-between p-1 mb-1 border-b">
      <div className="flex flex-row gap-1">
        <div className="border-r">
          <SquareBtn onClick={onShuffle} label="ðŸªƒ" ariaLabel="Shuffle" />
        </div>
      </div>
      <WordsBtn
        hoverMode={isByWrongIrregularWords}
        onClick={onWrongWords}
        current={finishedIrregularWords.length}
        wrong={wrongIrregularWords.length}
        total={allWords.length}
      />
    </div>
  )
}
