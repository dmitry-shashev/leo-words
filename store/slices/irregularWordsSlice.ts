import { IrregularWord } from '@/models/irregular-word'
import {
  IrregularWordStep,
  IrregularWordStepValue,
} from '@/types/irregular-word-step'
import {
  createSelector,
  createSlice,
  PayloadAction,
  Slice,
} from '@reduxjs/toolkit'
import { RootState } from '@/store/store'
import {
  getNextCircledValue,
  getPrevCircledValue,
} from '@/utils/getCircledValue'

function getEmptyIrregularWord(): IrregularWord {
  return {
    id: 0,
    infinitive: '',
    pastSimple: '',
    pastParticiple: '',
  }
}

interface IrregularWordsState {
  allIrregularWords: ReadonlyArray<IrregularWord>
  finishedIrregularWords: Array<IrregularWord>
  wrongIrregularWords: Array<IrregularWord>
  currentIrregularWord: IrregularWord
  savedCurrentIrregularWord: IrregularWord
  irregularWordStep: IrregularWordStepValue
  isByWrongIrregularWords: boolean
}

const initialState: IrregularWordsState = {
  allIrregularWords: [],
  finishedIrregularWords: [],
  wrongIrregularWords: [],
  currentIrregularWord: getEmptyIrregularWord(),
  savedCurrentIrregularWord: getEmptyIrregularWord(),
  irregularWordStep: IrregularWordStep.EMPTY,
  isByWrongIrregularWords: false,
}

export const irregularWordsSlice: Slice<
  IrregularWordsState,
  {
    setAllIrregularWords(
      state: IrregularWordsState,
      action: PayloadAction<Array<IrregularWord>>
    ): void
    goToNextIrregularWordStep(state: IrregularWordsState): void
    checkIrregularOk(state: IrregularWordsState): void
    checkIrregularNo(state: IrregularWordsState): void
    goIrregularBack(state: IrregularWordsState): void
    toggleByWrongIrregularWords(state: IrregularWordsState): void
  }
> = createSlice({
  name: 'irregularWordsSlice',
  initialState,
  reducers: {
    //----------------------------------------------------------------
    setAllIrregularWords(state, action) {
      state.allIrregularWords = action.payload
      state.finishedIrregularWords = []
      state.irregularWordStep = IrregularWordStep.CHECK_INFINITIVE
      state.currentIrregularWord = action.payload[0]!
    },
    //----------------------------------------------------------------
    goToNextIrregularWordStep(state: IrregularWordsState) {
      switch (state.irregularWordStep.value) {
        case IrregularWordStep.CHECK_INFINITIVE.value:
          state.irregularWordStep = IrregularWordStep.CHECK_PAST_SIMPLE
          break
        case IrregularWordStep.CHECK_PAST_SIMPLE.value:
          state.irregularWordStep = IrregularWordStep.CHECK_PAST_PARTICIPLE
          break
      }
    },
    //----------------------------------------------------------------
    checkIrregularOk(state: IrregularWordsState) {
      state.irregularWordStep = IrregularWordStep.CHECK_INFINITIVE

      if (state.isByWrongIrregularWords) {
        if (state.wrongIrregularWords.length === 1) {
          state.isByWrongIrregularWords = false
          state.wrongIrregularWords = []
          if (
            state.finishedIrregularWords.length >=
            state.allIrregularWords.length - 1
          ) {
            state.currentIrregularWord = state.allIrregularWords[0]!
            state.finishedIrregularWords = []
          } else {
            state.currentIrregularWord = state.savedCurrentIrregularWord
          }
          return
        }

        const { value } = getNextCircledValue(
          state.currentIrregularWord,
          state.wrongIrregularWords
        )
        state.wrongIrregularWords = state.wrongIrregularWords.filter(
          (v) => v.id !== state.currentIrregularWord.id
        )
        state.currentIrregularWord = value
        return
      }

      state.finishedIrregularWords.push(state.currentIrregularWord)
      const { value, reset } = getNextCircledValue(
        state.currentIrregularWord,
        state.allIrregularWords
      )
      state.currentIrregularWord = value
      if (reset) {
        if (state.wrongIrregularWords.length > 0) {
          state.isByWrongIrregularWords = true
          state.savedCurrentIrregularWord = state.currentIrregularWord
          state.currentIrregularWord = state.wrongIrregularWords[0]!
          return
        }
        state.finishedIrregularWords = []
      }
    },
    //----------------------------------------------------------------
    checkIrregularNo(state: IrregularWordsState) {
      state.irregularWordStep = IrregularWordStep.CHECK_INFINITIVE

      if (state.isByWrongIrregularWords) {
        const { value } = getNextCircledValue(
          state.currentIrregularWord,
          state.wrongIrregularWords
        )
        state.currentIrregularWord = value
        return
      }

      state.wrongIrregularWords.push(state.currentIrregularWord)
      state.finishedIrregularWords.push(state.currentIrregularWord)
      const { value, reset } = getNextCircledValue(
        state.currentIrregularWord,
        state.allIrregularWords
      )
      state.currentIrregularWord = value
      if (reset) {
        if (state.wrongIrregularWords.length > 0) {
          state.isByWrongIrregularWords = true
          state.savedCurrentIrregularWord = state.currentIrregularWord
          state.currentIrregularWord = state.wrongIrregularWords[0]!
          return
        }
        state.finishedIrregularWords = []
      }
    },
    //----------------------------------------------------------------
    goIrregularBack(state: IrregularWordsState) {
      const { value } = getPrevCircledValue(
        state.currentIrregularWord,
        state.allIrregularWords
      )

      state.irregularWordStep = IrregularWordStep.CHECK_INFINITIVE
      state.currentIrregularWord = value
      state.finishedIrregularWords = state.finishedIrregularWords.filter(
        (v) => v.id !== state.currentIrregularWord.id
      )
      state.wrongIrregularWords = state.wrongIrregularWords.filter(
        (v) => v.id !== state.currentIrregularWord.id
      )
    },
    //----------------------------------------------------------------
    toggleByWrongIrregularWords(state) {
      if (state.wrongIrregularWords.length === 0) {
        return
      }
      const value = !state.isByWrongIrregularWords
      if (value) {
        state.savedCurrentIrregularWord = state.currentIrregularWord
        state.currentIrregularWord = state.wrongIrregularWords[0]!
      } else {
        state.currentIrregularWord = state.savedCurrentIrregularWord
      }
      state.isByWrongIrregularWords = value
    },
    //----------------------------------------------------------------
  },
})

export const {
  setAllIrregularWords,
  goToNextIrregularWordStep,
  toggleByWrongIrregularWords,
  checkIrregularOk,
  checkIrregularNo,
  goIrregularBack,
} = irregularWordsSlice.actions

//----------------------------------------------------------------

export const getIrregularWordsState = (
  rootState: RootState
): IrregularWordsState => rootState.irregularWordsReducer

export const getAllIrregularWords = createSelector(
  getIrregularWordsState,
  (state) => state.allIrregularWords
)

export const getIrregularWordStep = createSelector(
  getIrregularWordsState,
  (state) => state.irregularWordStep
)

export const getIsFinalIrregularWordsStep = createSelector(
  getIrregularWordStep,
  (step) => step.value === IrregularWordStep.CHECK_PAST_PARTICIPLE.value
)

export const getCurrentIrregularWord = createSelector(
  getIrregularWordsState,
  (state) => state.currentIrregularWord
)

export const getIsByIrregularWrongWords = createSelector(
  getIrregularWordsState,
  (state) => state.isByWrongIrregularWords
)

export const getFinishedIrregularWords = createSelector(
  getIrregularWordsState,
  (state) => state.finishedIrregularWords
)

export const getWrongIrregularWords = createSelector(
  getIrregularWordsState,
  (state) => state.wrongIrregularWords
)

//----------------------------------------------------------------

export default irregularWordsSlice.reducer
