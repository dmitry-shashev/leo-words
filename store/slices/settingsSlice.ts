import { Sound, SoundValue } from '@/types/sound'
import { Lang, LangValue } from '@/types/lang'
import { createSelector, createSlice, Slice } from '@reduxjs/toolkit'
import { RootState } from '@/store/store'

interface SettingsState {
  sound: SoundValue
  lang: LangValue
}

const initialState: SettingsState = {
  sound: Sound.ON,
  lang: Lang.EN,
}

export const settingsSlice: Slice<
  SettingsState,
  {
    toggleSound(state: SettingsState): void
    toggleLang(state: SettingsState): void
  }
> = createSlice({
  name: 'settingsSlice',
  initialState,
  reducers: {
    toggleSound(state) {
      state.sound = state.sound.value === Sound.ON.value ? Sound.OFF : Sound.ON
    },
    toggleLang(state) {
      state.lang = state.lang.value === Lang.EN.value ? Lang.RU : Lang.EN
    },
  },
})

export const { toggleSound, toggleLang } = settingsSlice.actions

//----------------------------------------------------------------

const getSettingsState = (rootState: RootState): SettingsState =>
  rootState.settingsReducer

export const getSettingsSound = createSelector(
  getSettingsState,
  (state) => state.sound
)

export const getSettingsLang = createSelector(
  getSettingsState,
  (state) => state.lang
)

export default settingsSlice.reducer
