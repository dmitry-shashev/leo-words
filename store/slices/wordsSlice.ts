import { Word } from '@/models/word'
import {
  createSelector,
  createSlice,
  PayloadAction,
  Slice,
} from '@reduxjs/toolkit'
import { RootState } from '@/store/store'
import { Lang } from '@/types/lang'
import {
  getNextCircledValue,
  getPrevCircledValue,
} from '@/utils/getCircledValue'
import { SpeedMode, SpeedModeValue } from '@/types/speed-mode'
import { getEmptyWord } from '@/store/store-utils'
import { getLastWords } from '@/utils/getLastWords'

interface CheckButtonAction {
  speedMode: SpeedModeValue
}

interface WordsState {
  allWords: ReadonlyArray<Word>
  finishedWords: Array<Word>
  wrongWords: Array<Word>
  currentWord: Word
  savedCurrentWord: Word
  isCheckState: boolean
  isByWrongWords: boolean
}

const initialState: WordsState = {
  allWords: [],
  currentWord: getEmptyWord(),
  savedCurrentWord: getEmptyWord(),
  finishedWords: [],
  wrongWords: [],
  isCheckState: false,
  isByWrongWords: false,
}

export const wordsSlice: Slice<
  WordsState,
  {
    setAllWords(state: WordsState, action: PayloadAction<Array<Word>>): void
    goToCheckState(state: WordsState): void
    checkOk(state: WordsState, action: PayloadAction<CheckButtonAction>): void
    checkNo(state: WordsState, action: PayloadAction<CheckButtonAction>): void
    goBack(state: WordsState, action: PayloadAction<CheckButtonAction>): void
    toggleByWrongWords(state: WordsState): void
  }
> = createSlice({
  name: 'wordsSlice',
  initialState,
  reducers: {
    //----------------------------------------------------------------
    setAllWords(state, action) {
      state.isCheckState = false
      state.isByWrongWords = false
      state.finishedWords = []
      state.wrongWords = []
      state.allWords = action.payload
      state.currentWord = action.payload[0]!
    },
    //----------------------------------------------------------------
    goToCheckState(state) {
      state.isCheckState = true
    },
    //----------------------------------------------------------------
    checkOk(state, action) {
      state.isCheckState = false

      if (state.isByWrongWords) {
        if (state.wrongWords.length === 1) {
          state.isByWrongWords = false
          state.wrongWords = []
          if (state.finishedWords.length >= state.allWords.length - 1) {
            state.currentWord = state.allWords[0]!
            state.finishedWords = []
          } else {
            state.currentWord = state.savedCurrentWord
          }
          return
        }

        const { value } = getNextCircledValue(
          state.currentWord,
          state.wrongWords
        )
        state.wrongWords = state.wrongWords.filter(
          (v) => v.id !== state.currentWord.id
        )
        state.currentWord = value
        return
      }

      state.finishedWords.push(state.currentWord)
      const { value, reset } = getNextCircledValue(
        state.currentWord,
        state.allWords
      )
      state.currentWord = value
      if (reset) {
        if (state.wrongWords.length > 0) {
          state.isByWrongWords = true
          state.savedCurrentWord = state.currentWord
          state.currentWord = state.wrongWords[0]!
          return
        }
        state.finishedWords = []
      }

      if (action.payload.speedMode.value === SpeedMode.YES.value) {
        state.isCheckState = true
      }
    },
    //----------------------------------------------------------------
    checkNo(state, action) {
      state.isCheckState = false

      if (state.isByWrongWords) {
        const { value } = getNextCircledValue(
          state.currentWord,
          state.wrongWords
        )
        state.currentWord = value
        return
      }

      state.wrongWords.push(state.currentWord)
      state.finishedWords.push(state.currentWord)
      const { value, reset } = getNextCircledValue(
        state.currentWord,
        state.allWords
      )
      state.currentWord = value
      if (reset) {
        if (state.wrongWords.length > 0) {
          state.isByWrongWords = true
          state.savedCurrentWord = state.currentWord
          state.currentWord = state.wrongWords[0]!
          return
        }
        state.finishedWords = []
      }

      if (action.payload.speedMode.value === SpeedMode.YES.value) {
        state.isCheckState = true
      }
    },
    //----------------------------------------------------------------
    goBack(state, action) {
      state.isCheckState = false
      const { value } = getPrevCircledValue(state.currentWord, state.allWords)
      state.currentWord = value
      state.finishedWords = state.finishedWords.filter(
        (v: Word) => v.id !== state.currentWord.id
      )
      state.wrongWords = state.wrongWords.filter(
        (v: Word) => v.id !== state.currentWord.id
      )

      if (action.payload.speedMode.value === SpeedMode.YES.value) {
        state.isCheckState = true
      }
    },
    //----------------------------------------------------------------
    toggleByWrongWords(state) {
      if (state.wrongWords.length === 0) {
        return
      }
      const value = !state.isByWrongWords
      if (value) {
        state.savedCurrentWord = state.currentWord
        state.currentWord = state.wrongWords[0]!
      } else {
        state.currentWord = state.savedCurrentWord
      }
      state.isByWrongWords = value
    },
  },
})

export const {
  goToCheckState,
  checkOk,
  checkNo,
  goBack,
  toggleByWrongWords,
  setAllWords,
} = wordsSlice.actions

//----------------------------------------------------------------

export const getWordsState = (rootState: RootState): WordsState =>
  rootState.wordsReducer

export const getWordsAll = createSelector(
  getWordsState,
  (state) => state.allWords
)

export const getLastWordsAmount = createSelector(
  getWordsAll,
  (allWords) => getLastWords(allWords).length
)

export const getWordsWrong = createSelector(
  getWordsState,
  (state) => state.wrongWords
)

export const getWordsFinished = createSelector(
  getWordsState,
  (state) => state.finishedWords
)

export const getCurrentWordValue = (rootState: RootState): string => {
  if (rootState.settingsReducer.lang.value === Lang.EN.value) {
    return rootState.wordsReducer.currentWord.wordValue
  }
  return rootState.wordsReducer.currentWord.combinedTranslation
}

export const getCurrentWordTranslation = (rootState: RootState): string => {
  if (rootState.settingsReducer.lang.value === Lang.EN.value) {
    return rootState.wordsReducer.currentWord.combinedTranslation
  }
  return rootState.wordsReducer.currentWord.wordValue
}

export const getCurrentWordPicture = createSelector(
  getWordsState,
  (state) => state.currentWord.picture
)

export const getCurrentWord = createSelector(
  getWordsState,
  (state) => state.currentWord
)

export const getCurrentWordId = createSelector(
  getCurrentWord,
  (state) => state.id
)

export const getIsCheckState = createSelector(
  getWordsState,
  (state) => state.isCheckState
)

export const getIsByWrongWords = createSelector(
  getWordsState,
  (state) => state.isByWrongWords
)

export const getNextPrev10Words = createSelector(getWordsState, (state) => {
  let result: Array<Word> = []
  let current: Word = state.currentWord
  let prev: Word = state.currentWord
  for (let i = 0; i < 10; ++i) {
    current = getNextCircledValue(current, state.allWords).value
    if (!result.find((v) => v.id === current.id)) {
      result.push(current)
    }
    prev = getPrevCircledValue(current, state.allWords).value

    if (!result.find((v) => v.id === prev.id)) {
      result.push(prev)
    }
  }
  return result
})

export default wordsSlice.reducer
