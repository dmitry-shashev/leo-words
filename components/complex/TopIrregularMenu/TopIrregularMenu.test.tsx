import { test, expect } from 'vitest'
import { createStore } from '@/store/store'
import { setAllWords } from '@/store/slices/wordsSlice'
import { testWords } from '@/components/test-data/test-words'
import { act, render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { TopIrregularMenu } from '@/components/complex/TopIrregularMenu/TopIrregularMenu'

test('TopIrregularMenu', () => {
  const store = createStore()
  store.dispatch(setAllWords(testWords))

  const { asFragment, getByLabelText } = render(
    <Provider store={store}>
      <TopIrregularMenu />
    </Provider>
  )
  const shuffleBtn = getByLabelText('Shuffle')

  function checkControl(control: HTMLElement): void {
    act(() => {
      control.click()
    })
    expect(asFragment()).toMatchSnapshot()
    act(() => {
      control.click()
    })
    expect(asFragment()).toMatchSnapshot()
  }

  checkControl(shuffleBtn)
})
