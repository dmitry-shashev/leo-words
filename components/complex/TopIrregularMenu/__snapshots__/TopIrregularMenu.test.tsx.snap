// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TopIrregularMenu 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row justify-between p-1 mb-1 border-b"
  >
    <div
      class="flex flex-row gap-1"
    >
      <div
        class="border-r"
      >
        <button
          aria-label="Shuffle"
          class="w-10 h-10 text-gray-400 text-xl opacity-30"
          type="button"
        >
          ðŸªƒ
        </button>
      </div>
    </div>
    <button
      aria-label="Words button"
      class=" h-10 px-2 text-xl"
      type="button"
    >
      <span
        aria-label="Counter current"
        class="text-emerald-500"
      >
        0
      </span>
      <span
        class="mx-1"
      >
        /
      </span>
      <span
        aria-label="Counter wrong"
        class="text-orange-700"
      >
        0
      </span>
      <span
        class="mx-1"
      >
        /
      </span>
      <span
        aria-label="Counter total"
        class="text-gray-400"
      >
        0
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`TopIrregularMenu 2`] = `
<DocumentFragment>
  <div
    class="flex flex-row justify-between p-1 mb-1 border-b"
  >
    <div
      class="flex flex-row gap-1"
    >
      <div
        class="border-r"
      >
        <button
          aria-label="Shuffle"
          class="w-10 h-10 text-gray-400 text-xl opacity-30"
          type="button"
        >
          ðŸªƒ
        </button>
      </div>
    </div>
    <button
      aria-label="Words button"
      class=" h-10 px-2 text-xl"
      type="button"
    >
      <span
        aria-label="Counter current"
        class="text-emerald-500"
      >
        0
      </span>
      <span
        class="mx-1"
      >
        /
      </span>
      <span
        aria-label="Counter wrong"
        class="text-orange-700"
      >
        0
      </span>
      <span
        class="mx-1"
      >
        /
      </span>
      <span
        aria-label="Counter total"
        class="text-gray-400"
      >
        0
      </span>
    </button>
  </div>
</DocumentFragment>
`;
